### ISSUE TITLE
setup.py incorrectly detects OpenMP compilation when used with single-threaded OpenBLAS

### ISSUE THREAD

**Issue by mczwier** [Amber-MD/pytraj#1238]
OpenBLAS compiled in single-threaded mode always contains a routine `openblas_get_num_threads()`, (always returning 1 for single-threaded compilation). This means that the check for a multithreaded libcpptraj on [line 163 of `setup.py`](https://github.com/Amber-MD/pytraj/blob/master/setup.py#L163) incorrectly determines that libcpptraj is compiled with OpenMP support when libcpptraj is linked against OpenBLAS, even when serial compilation of cpptraj is requested (and performed). This currently prevents out-of-the-box Amber16 compilation when linked against OpenBLAS version 0.2.18 on Linux.

A possible fix is to change [line 163 of `setup.py`](https://github.com/Amber-MD/pytraj/blob/master/setup.py#L163) to read
```
libcpptraj_has_openmp = ([line for line in output_openmp_check 
                          if ('get_num_threads' in line.lower() 
                              and not 'openblas_get_num_threads' in line.lower()
                             )]  != [])
```

**Comment by hainm** [comment]
hi, thanks for your suggestion. I will have this fixed in github after trying your suggestion. cheers.

**Comment by hainm** [comment]
Q: what if both OpenMP and libcpptraj were compiled with openmp?

In that case
```python
('get_num_threads' in line.lower() 
      and not 'openblas_get_num_threads' in line.lower()
)
```
will return False

**Comment by mczwier** [comment]
That's indeed a concern. Maybe a check for `omp_num_threads` instead of `get_num_threads`, and then omit the secondary `openblas_get_num_threads` check?

**Comment by hainm** [comment]
I just tried with openblas (0.2.14) in Anaconda distribution.

`libcpptraj.so` does not show any sign of `get_num_threads` but `libopenblas` does

```python
cpptraj/lib (master) $ nm libcpptraj.so |grep thread
                 w pthread_cancel
00000000000b2800 W pthread_equal
```` 

```python
$ nm /home/haichit/anaconda3//lib/libopenblas.so.0 |grep thread
0000000000174300 T goto_set_num_threads
0000000000174320 T openblas_get_num_threads
0000000000174a70 T openblas_get_num_threads_
0000000000174a50 T openblas_set_num_threads
0000000000174a60 T openblas_set_num_threads_
```

To verify that my `libcpptraj.so` is compiled with openblas, I open it by vim and saw this line
`[hai's place holder] bfft_back_1_^@pubfft_forward_1_^@pubfft_init_1_^@libopenblas.so.0^@libgfortran.so.3^@libnetcdf.so```

I will try your version (0.2.18) so see the difference.

@drroe: ping you here in case you have any comment.

**Comment by mczwier** [comment]
I did a static link against `libopenblas`, as in `GOTO=/path/to/libopenblas.a` when running `configure`, so everything (used) in `libopenblas.a` wound up in `libcpptraj.so`.

**Comment by hainm** [comment]
I see now.

May be we can check 
```python
$ ldd libcpptraj.so |grep libgomp
	libgomp.so.1 => /usr/lib64/libgomp.so.1 (0x00002af2cf29b000)
```

what do you think about the reliability of that on Linux?

Need to use `otool -L` for mac.

**Comment by mczwier** [comment]
Would that work for compilers other than gfortran?

**Comment by hainm** [comment]
uhm I am naive. :D

**Comment by mczwier** [comment]
Hmm...the intel libraries appear to contain `__kmp_omp_num_threads` and `omp_num_threads` symbols. At least for intel and gfortran, testing whether the output of `nm libcpptraj.so` contains `omp_num_threads` should work. Can't test this for the PGI compilers, though a quick Google seems to indicate that it should work for PGI as well.

**Comment by hainm** [comment]
you meant `omp_num_threads` does not exists in libopenblas.a?

**Comment by hainm** [comment]
`omp_num_threads` does not work for my case

```python
$ nm libcpptraj.so |grep omp_num_threads
haichit@casegroup ~/pytraj_git/pytraj/cpptraj/lib (master) $ nm libcpptraj.so |grep threa
                 U omp_get_num_threads@@OMP_1.0
                 U omp_get_thread_num@@OMP_1.0
                 w pthread_cancel@@GLIBC_2.2.5
00000000000b2d90 W pthread_equal
```

**Comment by hainm** [comment]
so may be check `omp_get_num_threads`?

**Comment by mczwier** [comment]
Ah, crap, yes.

**Comment by hainm** [comment]
cool.

**Comment by mczwier** [comment]
Holy crap, that's the fastest turnaround on a fix I've ever seen. Thanks!

**Comment by hainm** [comment]
Holy crap, I am happy to have that fix. thanks for bring this to us. cheers.

**Comment by hainm** [comment]
close this issued with given fix on #1239.
